### PR TITLE
[Coverage] Consolidate visitation logic for functions and nominal types

### PR DESCRIPTION
**Pull request by vedantk**
This should address an ASAN failure which arose due to the PGOMapping
ASTWalker not being updated in sync with the other profiling-related
walkers.

rdar://39534066

### CODE REVIEW

**Comment by vedantk** [comment]
@swift-ci please smoke test and merge

File: lib/SIL/SILProfiler.cpp [hunk 1]
```diff
@@ -768,26 +795,21 @@ struct CoverageMapping : public ASTWalker {
...
-      return continueWalk;
+      });
     } else if (auto *TLCD = dyn_cast<TopLevelCodeDecl>(D)) {
       assignCounter(TLCD->getBody());
       ImplicitTopLevelBody = TLCD->getBody();
     } else if (auto *NTD = dyn_cast<NominalTypeDecl>(D)) {
-      bool continueWalk = Parent.isNull();
-      if (continueWalk) {
+      return visitNominalTypeDecl(*this, NTD, [&] {
         ParentNominalType = NTD;
```
Comment (slavapestov) [inline lib/SIL/SILProfiler.cpp]: What happens when you have nested types?

Comment (vedantk) [reply lib/SIL/SILProfiler.cpp]: When a constructor for a type is emitted in SILGen, we lazily create a SILProfiler instance for the nominal type associated with the constructor. The same holds for nested types.

Comment (vedantk) [reply lib/SIL/SILProfiler.cpp]: Er, that could've been clearer, sorry. We'd create a new ASTWalker for the nested type, where the nested nominal-type-decl would be the root of the walk.

**Comment by vedantk** [comment]
@shahmishal I'm seeing a strange failure on Linux:

> https://ci.swift.org/job/swift-PR-Linux-smoke-test/6339
> [xUnit] [ERROR] - Test reports were found but not all of them are new. Did all the tests run?
>   * /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/lit-tests.xml is 1 mo 26 days old

Any idea what this is about?

**Comment by vedantk** [comment]
@swift-ci Please clean smoke test Linux platform

**Comment by shahmishal** [comment]
real failure: 

```
12:09:40 CMake Error at CMakeLists.txt:103 (message):
12:09:40   LLDB test compilers not specified.  Tests will not run
12:09:40 
12:09:40 
12:09:40 -- Configuring incomplete, errors occurred!
```

**Comment by vedantk** [comment]
@shahmishal Thanks, I'm tracking a fix in rdar://39537655. I think I've gotten to the bottom of it..

**Comment by vedantk** [comment]
@shahmishal Heh, today's just not my day. The Linux job failed with a Jenkins exception this time around:

> https://ci.swift.org/job/swift-PR-Linux-smoke-test/6344
> 15:38:11 [177/334] CompileC: CoreFoundation/String.subproj/CFCharacterSet.c
> 15:45:44 FATAL: command execution failed
> 15:45:44 
> java.io.EOFException
> 15:45:44 	at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2624)

I've got a fix for the lldb cmake issue, though: https://github.com/apple/swift-lldb/pull/541

**Comment by vedantk** [comment]
@swift-ci Please clean smoke test Linux platform

**Pull request merged**